packages/babel-plugin-proposal-partial-application/src/index.js

```diff
@@ -0,0 +1,178 @@
+import { declare } from "@babel/helper-plugin-utils";
+import syntaxPartialApplication from "@babel/plugin-syntax-partial-application";
+import { types as t } from "@babel/core";
+
+export default declare(api => {
+  api.assertVersion(7);
+
+  /**
+   * a function to figure out if a call expression has
+   * ArgumentPlaceholder as one of its arguments
+   * @param node a callExpression node
+   * @returns boolean
+   */
+  function hasArgumentPlaceholder(node) {
+    return node.arguments.some(arg => t.isArgumentPlaceholder(arg));
+  }
+
+  /**
+   * Unwrap the arguments of a CallExpression, and creates
+   * assignmentExpression from them and returns them as an array.
+   * @param {Object} node CallExpression node
+   * @returns {Array<AssignmentExpression>}
+   */
+  function unwrapArguments(node, scope) {
+    let newNode;
+    let argId;
+    return node.arguments.reduce((acc, arg) => {
+      if (!t.isArgumentPlaceholder(arg) && !t.isNumericLiteral(arg)) {
+        argId = scope.generateUidIdentifier("_param");
+        scope.push({ id: argId });
+        newNode = t.assignmentExpression("=", t.cloneNode(argId), arg);
+        acc.push(newNode);
+      }
+      return acc;
+    }, []);
+  }
+
+  /**
+   * Unwraps all of the arguments in a CallExpression
+   * and replaces ArgumentPlaceholder type with Identifier
+   * and gives them a unique name.
+   * @param {Object} node
+   * @param {Object} scope
+   * @returns {Array<Expression>} the arguments
+   */
+  function unwrapAllArguments(node, scope) {
+    const clone = t.cloneNode(node);
+
+    return clone.arguments.map(arg => {
+      if (t.isArgumentPlaceholder(arg)) {
+        return Object.assign({}, arg, {
+          type: "Identifier",
+          name: scope.generateUid("_argPlaceholder"),
+        });
+      }
+      return arg;
+    });
+  }
+
+  /**
+   * It replaces the values of non-placeholder args in allArgs
+   * @param {Array<Node>} nonPlaceholderArgs that has no placeholder in them
+   * @param {Array<Node>} args
+   */
+  function mapNonPlaceholderToLVal(nonPlaceholderArgs, allArgsList) {
+    const clonedArgs = Array.from(allArgsList);
```

Comment:
Why do you need to clone this array? You are only cloning the array, not its elements.

Reply:
you are correct, this is useless. I removed it.